Felfel.Logging/ExceptionParser.cs

```diff
@@ -0,0 +1,313 @@
+            var flattened = aggregateException.Flatten();
+            foreach (var innerException in flattened.InnerExceptions)
+            {
+                Parse(exceptionData, innerException, builder, indent + 1);
+            }
+        }
+
+        /// <summary>
+        /// Indents the current line.
+        /// </summary>
+        internal static void Indent(StringBuilder builder, int count)
+        {
+            builder.Append("".PadLeft(count * 4));
+        }
+
+        /// <summary>
+        /// Appends text with optional indent.
+        /// </summary>
+        internal static void Append(StringBuilder builder, int indent, string value)
+        {
+            var paragraphs = value.Split(new[] { "\n", "\r\n" }, StringSplitOptions.None);
+            for (int index = 0; index < paragraphs.Length; index++)
+            {
+                var paragraph = paragraphs[index];
+                Indent(builder, indent);
+
+                builder.Append(paragraph);
+
+                //omit the line break after the last paragraph
+                if (index < paragraphs.Length - 1) builder.AppendLine();
+            }
+        }
+
+        /// <summary>
+        /// Appends an optionally indented line.
+        /// </summary>
+        internal static void AppendLine(StringBuilder builder, int indent, string value)
+        {
+            Append(builder, indent, value);
+            builder.AppendLine();
+        }
+
+        /// <summary>
+        /// Appends a formatted string with optional indent.
+        /// </summary>
+        internal static void AppendFormat(StringBuilder builder, int indent, string value, params object[] args)
+        {
+            var formatted = String.Format(value, args);
+            Append(builder, indent, formatted);
+        }
+
+        /// <summary>
+        /// Gets a hash built based on all exceptions in the submitted <see cref="ExceptionData"/>
+        /// object. Hidden exceptions result in multiple dotted hashes, to enable comparison of
+        /// other hashes that ultimately result in the same exception.<br/>
+        /// Note that if more than 10 exceptions occur, the last hash represents multiple exceptions
+        /// (10th and higher).
+        /// </summary>
+        /// <returns></returns>
+        internal static string CalculateHash(ExceptionData exceptionData)
+        {
+            var list = new List<Exception>(exceptionData.HiddenExceptions);
+            list.Insert(0, exceptionData.RootException);
+
+            using (MD5 md5 = MD5.Create())
+            {
+                List<string> sourceStrings = list.Select(e => GetStackTraces(e)).ToList();
+
+                //TBD: this routine would just create a single (32 character) hash of all exceptions:
+                //var unhashed = Encoding.Unicode.GetBytes(String.Concat(sourceStrings));
+                //var hash = md5.ComputeHash(unhashed);
+                //return String.Concat(hash.Select(b => b.ToString("X2")));
```

Comment:
Dislaimer: I wrote that particular class years ago, so to be taken with a grain of salt: I think the justification was that me might have been able to identify exceptions occurring in different locations (e.g. exception X in a helper method, which is used by A / B). This would for example allow you to easily discover issues in packages that you are sharing across components and/or applications.